### PR TITLE
[WIP] Prototype classic retriever

### PR DESCRIPTION
**Pull request by jdconrad**
Re-opening this change against a feature branch.

https://github.com/elastic/elasticsearch/pull/100561

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 1]
```diff
@@ -1592,6 +1601,9 @@ private SearchSourceBuilder parseXContent(XContentParser parser, boolean checkTr
                 throw new ParsingException(parser.getTokenLocation(), "Unexpected token [" + token + "] found after the main object.");
             }
         }
+        if (retrieverBuilder != null) {
+            retrieverBuilder.extractToSearchSourceBuilder(this);
+        }
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1606]: FYI, I am not sure about this. It seems to be SearchSourceBuilder should have a bunch of retrievers it runs. Extracting like this seems like an anti-pattern.

Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1606]: I agree with your thoughts here, and this is not intended to be the end design. I'm trying to chunk off as little as possible, so it made sense to me to do it this way as the first step to get the parsing code in. Next step I'll make it so the `SearchSourceBuilder` is rebuilt as a retriever and that is passed to the shard.

**Comment by jdconrad** [comment]
After speaking with @mayya-sharipova, I'm going to push this to the feature branch as-is, so we can more easily work in parallel together.

**Comment by Rassyan** [comment]
Just checking, does this resolve #88921?

**Pull request merged**